src/jsonpatcherproxy.js

```diff
@@ -76,15 +107,29 @@ var JSONPatcherProxy = (function() {
         return Reflect.get(target, propKey, receiver);
       },
       set: function(target, key, receiver) {
-        const distPath =
-          path + '/' + JSONPatcherProxy.escapePathComponent(key.toString());
+
+        /* each proxified object has its path cached, we need to use that instead of `path` variable
+        because at one point in the future, paths might change and we will simply update our cache instead of 
+        proxifying again.  */
+
+        var destinationPropKey = instance.getOrSetObjectPath(target, path) + '/' + JSONPatcherProxy.escapePathComponent(key);
+
+        /* in case the set value is already proxified by a different instance of JSONPatcherProxy */
```

Comment:
>Doesn't it also match the case when it is proxified by the same instance, but is just crosss-referenced? 

No. Because there is an additional condition in [this line](https://github.com/Palindrom/JSONPatcherProxy/blob/Fix-nested-array-deletion/src/jsonpatcherproxy.js#L121).

It will only deep clone if the proxy is coming from another instance (doesn't exist in this instance's map).